### PR TITLE
feat: add filtering

### PR DESCRIPTION
**Pull request by samsja**
Adding pre-filtering for Qdrant indexer : https://github.com/jina-ai/docarray/issues/356

Need to merge https://github.com/jina-ai/docarray/pull/378 first

### CODE REVIEW

**Review by alaeddine-13 — COMMENTED** [review]
Please also pin the docarray version in requirements.txt since this depends on a new version in docarray

File: tests/integration/test_integration.py [hunk 1]
```diff
@@ -41,3 +46,48 @@ def test_reload_keep_state(docker_compose):
...
         second_search = f.search(inputs=docs)
         assert len(first_search[0].matches) == len(second_search[0].matches)
+
+
+numeric_operators_qdrant = {
+    'gte': operator.ge,
+    'gt': operator.gt,
+    'lte': operator.le,
+    'lt': operator.lt,
+}
```
Comment (bwanglzu) [inline tests/integration/test_integration.py:56]: i would make this more general rather than copy-paste everywhere

File: executor.py [hunk 2]
```diff
@@ -36,6 +37,7 @@ def __init__(
         :param serialize_config: DocumentArray serialize configuration.
         :param m: The maximum number of connections per element in all layers. Defaults to the default
             `m` in the Qdrant server.
+        :param columns: precise columns for the Indexer.
```
Comment (alaeddine-13) [inline executor.py:40]:
```suggestion
        :param columns: precise columns for the Indexer (used for filtering).
```

File: README.md [hunk 3]
```diff
@@ -100,4 +100,33 @@ with f:
...
+        'collection_name': 'test',
+        'n_dim': 3,
+        'columns': [('price', 'float')],
+    },
+)
+```
+
+Then you can pass a filter as a parameters when searching for document:
+```python
+filter = {'must': [{'key': 'price', 'range': {'gte': 30}}]}
```
Comment (bwanglzu) [inline README.md:125]: this is a reserved keyword in python, maybe rename it to `filters`

**Review by bwanglzu — CHANGES_REQUESTED** [review]
some comments

**Review by JoanFM — APPROVED** [review]

File: executor.py [hunk 4]
```diff
@@ -68,13 +71,14 @@ def index(self, docs: DocumentArray, **kwargs):
     def search(
         self,
         docs: 'DocumentArray',
+        parameters: Dict = {},
```
Comment (alaeddine-13) [inline executor.py:74]: maybe also document the parameter ?

File: tests/integration/test_integration.py [hunk 5]
```diff
@@ -41,3 +46,48 @@ def test_reload_keep_state(docker_compose):
...
+    with f:
+        f.index(docs)
+
+        for threshold in [10, 20, 30]:
+
+            filter = {'must': [{'key': 'price', 'range': {operator: threshold}}]}
+            doc_query = DocumentArray([Document(embedding=np.random.rand(n_dim))])
+            indexed_docs = f.search(doc_query, parameters={'filter': filter})
+
+            print(indexed_docs[0].matches)
```
Comment (alaeddine-13) [inline tests/integration/test_integration.py:87]: remove this

File: tests/integration/test_integration.py [hunk 6]
```diff
@@ -41,3 +46,48 @@ def test_reload_keep_state(docker_compose):
...
+            assert all(
```
Comment (alaeddine-13) [inline tests/integration/test_integration.py:88]: assert len here

File: tests/unit/test_executor.py [hunk 7]
```diff
@@ -144,3 +145,59 @@ def test_clear(docs, docker_compose):
...
+    indexer.index(docs)
+
+    for threshold in [10, 20, 30]:
+
+        filter = {'must': [{'key': 'price', 'range': {operator: threshold}}]}
+
+        doc_query = DocumentArray([Document(embedding=np.random.rand(n_dim))])
+        indexer.search(doc_query, parameters={'filter': filter})
+
+        print(doc_query[0].matches[:,'tags__price'])
```
Comment (alaeddine-13) [inline tests/unit/test_executor.py:195]: remove

File: tests/unit/test_executor.py [hunk 8]
```diff
@@ -144,3 +145,59 @@ def test_clear(docs, docker_compose):
...
+
+        assert all(
```
Comment (alaeddine-13) [inline tests/unit/test_executor.py:197]: assert len

**Review by bwanglzu — APPROVED** [review]
LGTM! btw why create a helper method instead of a fixture?

File: executor.py (hunk 4 above)
Comment (alaeddine-13) [reply executor.py:74]: and also say what is the parameter filter and you can set it, where to get documentation about it, ...

Comment (alaeddine-13) [reply executor.py:74]: because these information in the docstring go to hub website and will be parsed for users

**Comment by samsja** [comment]
> 

The issue is that the dictionary is used to parametrize another test. So it can't be a fixture ...
See here : https://github.com/jina-ai/executor-qdrant-indexer/blob/ed869f23618fdf55e7a1b655b74a6932e6bb8b25/tests/integration/test_integration.py#L50

File: executor.py (hunk 4 above)
Comment (samsja) [reply executor.py:74]: Docstring of the Executor endpoint goes on hub ?
I should not precise the filter here as this is general parameters. It contains things from core as well like `runtine args` etcc.

Rather I documented how to use filter in the README.md https://github.com/jina-ai/executor-qdrant-indexer/tree/feat-pre-filtering#vector-search

**Pull request merged**